src/_pytest/monkeypatch.py

```diff
@@ -271,6 +271,19 @@ def syspath_prepend(self, path):
         # https://github.com/pypa/setuptools/blob/d8b901bc/docs/pkg_resources.txt#L162-L171
         fixup_namespace_packages(str(path))
 
+        # A call to syspathinsert() usually means that the caller wants to
+        # import some dynamically created files, thus with python3 we
+        # invalidate its import caches.
+        # This is especially important when any namespace package is in used,
+        # since then the mtime based FileFinder cache (that gets created in
+        # this case already) gets not invalidated when writing the new files
+        # quickly afterwards.
+        try:
+            from importlib import invalidate_caches
+        except ImportError:
+            return
+        invalidate_caches()
```

Comment:
to make this more future-proof (in case we add code after this) I'd do this:

```
if sys.version_info >= (...):
    from importlib import invalidate_caches
    invalidate_caches()
```

this avoids general frustration that can be caused by backport modules which aren't intended to be run in this way

it also gives us good documentation on when we can remove the version-specific hax

Reply:
Good point.
I've just moved this though basically, and would not know the version to use without checking.

Reply:
fortunately the python docs are very helpful here

https://docs.python.org/3/library/importlib.html#importlib.invalidate_caches

> New in version 3.3.

Reply:
Yeah, let's add it in features after this is merged - no functional changes on master.